Support for overloaded functions

I saw a crash without them:

```
e: org.jetbrains.kotlin.backend.common.BackendException: Backend Internal error: Exception during IR lowering
File being compiled: /data/app/kochiku-worker/shared/build-partition/squareup/cash-treehouse/treehouse/apis/flows/src/commonMain/kotlin/com/squareup/cash/treehouse/flows/FlowsApp.kt
The root cause java.lang.RuntimeException was thrown at: org.jetbrains.kotlin.backend.jvm.codegen.FunctionCodegen.generate(FunctionCodegen.kt:47)
	at org.jetbrains.kotlin.backend.common.CodegenUtil.reportBackendException(CodegenUtil.kt:239)
	at org.jetbrains.kotlin.backend.common.CodegenUtil.reportBackendException$default(CodegenUtil.kt:235)
Caused by: java.lang.IllegalArgumentException: No argument for parameter VALUE_PARAMETER name:parameters index:2 type:@[Contextual] okio.ByteString?:
CALL 'public open fun getScreen (path: kotlin.String, navigator: com.squareup.cash.treehouse.navigation.Navigator, parameters: @[Contextual] okio.ByteString?): app.cash.redwood.treehouse.ZiplineTreehouseUi
```

--- a/zipline-kotlin-plugin/src/main/kotlin/app/cash/zipline/kotlin/AdapterGenerator.kt
+++ b/zipline-kotlin-plugin/src/main/kotlin/app/cash/zipline/kotlin/AdapterGenerator.kt
@@ -694,7 +694,7 @@ internal class AdapterGenerator(
     )
     outboundServiceClass.declarations += callHandlerProperty
 
-    for ((i, overridesList) in bridgedInterface.bridgedFunctionsWithOverrides.values.withIndex()) {
+    for ((i, overridesList) in bridgedInterface.bridgedFunctionsWithOverrides.withIndex()) {
       outboundServiceClass.irBridgedFunction(
         functionIndex = i,
         bridgedInterface = bridgedInterface,

--- a/zipline-kotlin-plugin/src/main/kotlin/app/cash/zipline/kotlin/BridgedInterface.kt
+++ b/zipline-kotlin-plugin/src/main/kotlin/app/cash/zipline/kotlin/BridgedInterface.kt
@@ -25,6 +25,7 @@ import org.jetbrains.kotlin.ir.builders.irGet
 import org.jetbrains.kotlin.ir.builders.irInt
 import org.jetbrains.kotlin.ir.builders.irTemporary
 import org.jetbrains.kotlin.ir.builders.irVararg
+import org.jetbrains.kotlin.ir.declarations.IrSimpleFunction
 import org.jetbrains.kotlin.ir.declarations.IrTypeParameter
 import org.jetbrains.kotlin.ir.declarations.IrValueParameter
 import org.jetbrains.kotlin.ir.declarations.IrVariable
@@ -72,35 +73,30 @@ internal class BridgedInterface(
 ) {
   val typeIrClass = classSymbol.owner
 
-  // TODO(jwilson): support overloaded functions?
-  val bridgedFunctionsWithOverrides: Map<Name, List<IrSimpleFunctionSymbol>>
+  val bridgedFunctionsWithOverrides: List<List<IrSimpleFunctionSymbol>>
     get() {
-      val result = mutableMapOf<Name, MutableList<IrSimpleFunctionSymbol>>()
+      val result = mutableMapOf<String, MutableList<IrSimpleFunctionSymbol>>()
       for (supertype in listOf(classSymbol.owner.defaultType) + classSymbol.owner.superTypes) {
         val supertypeClass = supertype.getClass() ?: continue
+        val functions = mutableListOf<IrSimpleFunction>()
         for (function in supertypeClass.functions) {
           if (function.name.identifier in NON_INTERFACE_FUNCTION_NAMES) continue
-          val overrides = result.getOrPut(function.name) { mutableListOf() }
-          overrides += function.symbol
+          functions += function
         }
         for (property in supertypeClass.properties) {
-          val getter = property.getter
-          if (getter != null) {
-            val overrides = result.getOrPut(getter.name) { mutableListOf() }
-            overrides += getter.symbol
-          }
-          val setter = property.setter
-          if (setter != null) {
-            val overrides = result.getOrPut(setter.name) { mutableListOf() }
-            overrides += setter.symbol
-          }
+          property.getter?.let { functions += it }
+          property.setter?.let { functions += it }
+        }
+        for (function in functions) {
+          val overrides = result.getOrPut(function.signature) { mutableListOf() }
+          overrides += function.symbol
         }
       }
-      return result
+      return result.values.toList()
     }
 
   val bridgedFunctions: List<IrSimpleFunctionSymbol>
-    get() = bridgedFunctionsWithOverrides.values.map { it[0] }
+    get() = bridgedFunctionsWithOverrides.map { it[0] }
 
   /** Declares local vars for all the serializers needed to bridge this interface. */
   fun declareSerializerTemporaries(

--- a/zipline-kotlin-plugin/tests/src/test/kotlin/app/cash/zipline/kotlin/ZiplineKotlinPluginTest.kt
+++ b/zipline-kotlin-plugin/tests/src/test/kotlin/app/cash/zipline/kotlin/ZiplineKotlinPluginTest.kt
@@ -369,6 +369,29 @@ class ZiplineKotlinPluginTest {
       .contains(":5:1 Only classes may implement ZiplineScoped, but " +
         "app.cash.zipline.testing.SomeInterface is an interface")
   }
+
+  @Test
+  fun `service has function with overloads`() {
+    val result = compile(
+      sourceFile = SourceFile.kotlin(
+        "main.kt",
+        """
+        package app.cash.zipline.testing
+
+        import app.cash.zipline.ZiplineService
+
+        interface ParentService : ZiplineService {
+            fun test(first: Int)
+        }
+
+        interface ChildService : ParentService, ZiplineService {
+            fun test(first: Int, second: Int)
+        }
+        """
+      )
+    )
+    assertEquals(KotlinCompilation.ExitCode.OK, result.exitCode, result.messages)
+  }
 }
 
 @OptIn(ExperimentalCompilerApi::class)

--- a/zipline/src/jniTest/kotlin/app/cash/zipline/ZiplineTest.kt
+++ b/zipline/src/jniTest/kotlin/app/cash/zipline/ZiplineTest.kt
@@ -49,10 +49,8 @@ class ZiplineTest {
     zipline.loadTestingJs()
   }
 
-  @After fun tearDown() {
-    runBlocking(dispatcher) {
-      zipline.close()
-    }
+  @After fun tearDown() = runBlocking(dispatcher) {
+    zipline.close()
   }
 
   @Test fun cannotTakeOrBindServiceAfterClose(): Unit = runBlocking(dispatcher) {

